### PR TITLE
CIV-13857 Lip Defendant response to application - upload documents

### PR DESCRIPTION
**Pull request by kalachandrasekar1**
### JIRA link ###

https://tools.hmcts.net/jira/browse/CIV-13857

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/views/features/generalApplication/respondent-hearing-preference.njk [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+{% from "../../macro/contact-us-for-help.njk" import contactUsForHelp %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+{% from "../../macro/page-title.njk" import setPageTitle %}
+{% from "../../macro/heading-with-caption.njk" import headingWithCaption %}
+
+{% block pageTitle %}
+  {{ setPageTitle(t, 'PAGES.GENERAL_APPLICATION.HEARING_PREFERENCE.TITLE') }}
+{% endblock %}
+
+{% block content %}
```
Comment (laarmada) [inline src/main/views/features/generalApplication/respondent-hearing-preference.njk:11]: please merge the latest of claim-details-tpl-dashboard.njk and refactor

File: src/main/views/features/generalApplication/respondent-upload-documents.njk [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+{% from "../../macro/heading-with-caption.njk" import headingWithCaption %}
+{% from "govuk/components/file-upload/macro.njk" import govukFileUpload %}
+{% from "../../macro/button.njk" import addButton %}
+{% from "../../macro/error-summary.njk" import errorSummary %}
+
+{% block pageTitle %}
+  {{ setPageTitle(t, 'PAGES.GENERAL_APPLICATION.RESPONDENT_UPLOAD_DOCUMENTS.TITLE') }}
+{% endblock %}
+
+{% block content %}
```
Comment (laarmada) [inline src/main/views/features/generalApplication/respondent-upload-documents.njk:15]: please merge the latest of claim-details-tpl-dashboard.njk and refactor

File: src/main/views/features/generalApplication/respondent-want-to-upload-documents.njk [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+{% from "../../macro/heading-with-caption.njk" import headingWithCaption %}
+{% from "../../macro/csrf.njk" import csrfProtection %}
+{% from "../../macro/page-title.njk" import setPageTitle %}
+{% from "../../macro/error-summary.njk" import errorSummary %}
+
+{% block pageTitle %}
+  {{ setPageTitle(t, 'PAGES.GENERAL_APPLICATION.RESPONDENT_WANT_TO_UPLOAD_DOC.PAGE_TITLE') }}
+{% endblock %}
+
+{% block content %}
```
Comment (laarmada) [inline src/main/views/features/generalApplication/respondent-want-to-upload-documents.njk:14]: please merge the latest of claim-details-tpl-dashboard.njk and refactor

File: src/main/services/features/generalApplication/response/respondentUploadEvidenceDocumentsService.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import {GenericForm} from 'form/models/genericForm';
+import {AppRequest} from 'models/AppRequest';
+import {TypeOfDocumentSectionMapper} from 'services/features/caseProgression/TypeOfDocumentSectionMapper';
+import config from 'config';
+import {CivilServiceClient} from 'client/civilServiceClient';
+import { FormValidationError } from 'form/validationErrors/formValidationError';
+import { t } from 'i18next';
+import {GaResponse} from 'models/generalApplication/response/response';
+
+const {Logger} = require('@hmcts/nodejs-logging');
```
Comment (laarmada) [inline src/main/services/features/generalApplication/response/respondentUploadEvidenceDocumentsService.ts:18]: where is the related test?

File: src/main/modules/nunjucks/index.ts [hunk 5]
```diff
@@ -104,7 +104,7 @@ export class Nunjucks {
         }
         if(key?.text)
           return ({...key, text: t(key?.text)});
-      });
+      }).filter(item => item);
```
Comment (laarmada) [inline src/main/modules/nunjucks/index.ts:107]: which is the reason of that?

**Review by laarmada — CHANGES_REQUESTED** [review]

File: src/main/routes/features/generalApplication/respondentUploadEvidenceDocmentsController.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    const claim = await getClaimById(claimId, req, true);
+    const redisKey = generateRedisKey(req);
+    const uploadEvidenceDocuments = new UploadGAFiles();
+    let form = new GenericForm(uploadEvidenceDocuments);
+    const formattedSummary = summarySection(
+      {
+        title: '',
+        summaryRows: [],
+      });
+    if (req?.session?.fileUpload) {
```
Comment (laarmada) [inline src/main/routes/features/generalApplication/respondentUploadEvidenceDocmentsController.ts:58]: is possible to extract to a function?

File: src/main/routes/features/generalApplication/respondentUploadEvidenceDocmentsController.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      const parsedData = JSON.parse(req?.session?.fileUpload);
+      form = new GenericForm(uploadEvidenceDocuments, parsedData);
+      req.session.fileUpload = undefined;
+    }
+    if (req.query?.id) {
```
Comment (laarmada) [inline src/main/routes/features/generalApplication/respondentUploadEvidenceDocmentsController.ts:63]: also this one.....

File: src/main/routes/features/generalApplication/respondentUploadEvidenceDocmentsController.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  } catch (error) {
+    next(error);
+  }
+}) as RequestHandler);
+
+respondentUploadEvidenceDocumentsController.post(GA_RESPONDENT_UPLOAD_DOCUMENT, upload.single('selectedFile'), (async (req: AppRequest, res: Response, next: NextFunction) => {
+  try {
+    const claimId = req.params.id;
+    const redisKey = generateRedisKey(req);
+    const claim: Claim = await getCaseDataFromStore(redisKey);
```
Comment (laarmada) [inline src/main/routes/features/generalApplication/respondentUploadEvidenceDocmentsController.ts:78]: why you dont use the same way as get: .... getClaimById

File: src/main/routes/features/generalApplication/respondentWantToUploadDocumentsController.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    if (form.hasErrors()) {
+      await  renderView(req, form, claim, claimId, res);
+    } else {
+      let redirectUrl;
+      if (req.body.option == YesNo.YES) {
+        redirectUrl = constructResponseUrlWithIdParams(claimId, GA_RESPONDENT_UPLOAD_DOCUMENT);
+      } else if (req.body.option == YesNo.NO) {
+        redirectUrl = 'test'; // TODO: add url
+      }
+      await saveRespondentWantToUploadDoc(redisKey, claim, req.body.option);
```
Comment (laarmada) [inline src/main/routes/features/generalApplication/respondentWantToUploadDocumentsController.ts:62]: you can pass the req and remove the line 49.... toy can include the generateRedisKey into the saveRespondentWantToUploadDoc method

**Review by laarmada — CHANGES_REQUESTED** [review]

**Review by laarmada — APPROVED** [review]

File: src/main/modules/nunjucks/index.ts (hunk 5 above)
Comment (laarmada) [reply src/main/modules/nunjucks/index.ts]: any idea of that? could you talk with Neeta the reason to avoid problems across other pages?

**Review by jeswanth-hmcts — DISMISSED** [review]

**Review by david-jeronimo — DISMISSED** [review]

File: src/main/services/features/generalApplication/response/respondentUploadEvidenceDocumentsService.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+      const errors = translateErrors(form.getAllErrors(), t);
+      req.session.fileUpload = JSON.stringify(errors);
+    }
+  } catch(error) {
+    logger.error(error);
+    throw error;
+  }
+};
+
+const translateErrors = (keys: FormValidationError[], t: (key: string) => string, formatValues?: { keyError: string, keyToReplace: string, valueToReplace: string }[]) => {
```
Comment (neeta-hmcts) [inline src/main/services/features/generalApplication/response/respondentUploadEvidenceDocumentsService.ts:78]: this is a duplicate method.. can be re-used from the application uploadEvidance service

File: src/main/views/features/generalApplication/respondent-upload-documents.njk [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+          {% if formattedSummary.summaryList.rows.length > 0 %}
+            <h2 class="govuk-heading-m">{{ t('PAGES.GENERAL_APPLICATION.UPLOAD_DOCUMENTS.LIST_TITLE') }}</h2>
+            {% for row in formattedSummary.summaryList.rows  %}
+              <dl class="govuk-summary-list">
+                <div class="govuk-summary-list__row">
+                    <dt class="govuk-summary-list__key">
+                      {{ row.key.text }}
+                    </dt>
+                  <dd class="govuk-summary-list__actions">
+                    <a class="govuk-link" href="{{ row.actions.items[0].href }}&action=REMOVE_DOC">
```
Comment (neeta-hmcts) [inline src/main/views/features/generalApplication/respondent-upload-documents.njk:55]: REMOVE_DOC action not required

**Review by neeta-hmcts — CHANGES_REQUESTED** [review]

File: src/main/views/features/generalApplication/respondent-upload-documents.njk (hunk 11 above)
Comment (kalachandrasekar1) [reply src/main/views/features/generalApplication/respondent-upload-documents.njk]: removed

File: src/main/services/features/generalApplication/response/respondentUploadEvidenceDocumentsService.ts (hunk 10 above)
Comment (kalachandrasekar1) [reply src/main/services/features/generalApplication/response/respondentUploadEvidenceDocumentsService.ts]: updated to reuse uploadEvidance service class method

**Review by neeta-hmcts — APPROVED** [review]

**Review by jeswanth-hmcts — DISMISSED** [review]

**Pull request merged**